### PR TITLE
feat: Set the user ip if SendDefaultPii is true.

### LINKED ISSUE #561: Send user.ip_address = {{ default }} when sendDefaultPii is true
If no explicit IpAddress was set by the user code (or framework integration) the SDK should set `{{ defaut }}` which tells the server to infer the IP Address to client connection as the `user.ip_address`.

### PR DESCRIPTION
**Pull request by lucas-zimerman**
Basically, what the title says, the server will set the user IP if SendDefaultPii is set.

Fixes #561

### CODE REVIEW

File: src/Sentry/Internal/MainSentryEventProcessor.cs [hunk 1]
```diff
@@ -42,6 +42,11 @@ internal class MainSentryEventProcessor : ISentryEventProcessor
         internal string? Release => _release.Value;
         internal Runtime? Runtime => _runtime.Value;
 
+        /// <summary>
+        /// A flag that tells the endpoint to figure out the user ip.
+        /// </summary>
+        internal string UserIpDefault = "{{auto}}";
```
Comment (bruno-garcia) [inline src/Sentry/Internal/MainSentryEventProcessor.cs:48]:
```suggestion
        internal string UserIpServerInferred = "{{auto}}";
```

File: src/Sentry/Internal/MainSentryEventProcessor.cs [hunk 2]
```diff
@@ -108,9 +113,11 @@ public SentryEvent Process(SentryEvent @event)
             }
 
             // Report local user if opt-in PII, no user was already set to event and feature not opted-out:
-            if (_options.SendDefaultPii && _options.IsEnvironmentUser && !@event.HasUser())
+            if (_options.SendDefaultPii && _options.IsEnvironmentUser)
```
Comment (bruno-garcia) [inline src/Sentry/Internal/MainSentryEventProcessor.cs:116]: `IsEnvironmentUser` this option control whether to use `Environment.UserName` or not. So I wouldn't guard your new option under this condition:


```csharp
if (_options.SendDefaultPii)
{
    if (_options.IsEnvironmentUser && !@event.HasUser())
    {
        @event.User.Username = Environment.UserName;
    }
    @event.User.IpAddress ??= UserIpServerInferred;
}

```

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs [hunk 3]
```diff
@@ -63,6 +63,43 @@ public void Process_SendDefaultPiiTrueIdEnvironmentTrue_UserNameSet()
             Assert.Equal(Environment.UserName, evt.User.Username);
         }
 
+        [Fact]
+        public void Process_SendDefaultPiiTrueIdEnvironmentTrueUserIpNull_UserIpSetDefaultIp()
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:67]:
```suggestion
        public void Process_SendDefaultPiiTrueAndUserIpNull_UserIpSetServerInferredIp()
```

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs [hunk 4]
```diff
@@ -63,6 +63,43 @@ public void Process_SendDefaultPiiTrueIdEnvironmentTrue_UserNameSet()
...
+        {
+            //Arrange
+            var evt = new SentryEvent();
+
+            _fixture.SentryOptions.SendDefaultPii = true;
+            _fixture.SentryOptions.IsEnvironmentUser = true;
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:73]:
```suggestion
```

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs [hunk 5]
```diff
@@ -63,6 +63,43 @@ public void Process_SendDefaultPiiTrueIdEnvironmentTrue_UserNameSet()
...
+
+            //Act
+            _ = sut.Process(evt);
+
+            //Assert
+            Assert.Equal(sut.UserIpDefault, evt.User.IpAddress);
+        }
+
+        [Fact]
+        public void Process_SendDefaultPiiTrueIdEnvironmentTrueUserIpSet_UserIpIgnoreDefaultIp()
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:84]:
```suggestion
        public void Process_SendDefaultPiiTrueAndUserIpSet_UserIpIgnoreServerInferredIp()
```

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs [hunk 6]
```diff
@@ -63,6 +63,43 @@ public void Process_SendDefaultPiiTrueIdEnvironmentTrue_UserNameSet()
...
+        {
+            //Arrange
+            var evt = new SentryEvent();
+            var @ip = "192.0.0.1";
+            evt.User.IpAddress = ip;
+
+            _fixture.SentryOptions.SendDefaultPii = true;
+            _fixture.SentryOptions.IsEnvironmentUser = true;
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:92]:
```suggestion
```

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs [hunk 7]
```diff
@@ -63,6 +63,43 @@ public void Process_SendDefaultPiiTrueIdEnvironmentTrue_UserNameSet()
...
+
+            var sut = _fixture.GetSut();
+
+            //Act
+            _ = sut.Process(evt);
+
+            //Assert
+            Assert.Equal(ip, evt.User.IpAddress);
+        }
+
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:102]:
```suggestion
        [Fact]
        public void Process_SendDefaultPiiFalse_UserIpNotSet()
        {
            //Arrange
            var evt = new SentryEvent();

            _fixture.SentryOptions.SendDefaultPii = false;
            var sut = _fixture.GetSut();

            //Act
            _ = sut.Process(evt);

            //Assert
            Assert.Null(evt.User.IpAddress);
        }
```

**Review by bruno-garcia — COMMENTED** [review]
Thanks for tackling this one @lucas-zimerman !

File: src/Sentry/Internal/MainSentryEventProcessor.cs (hunk 2 above)
Comment (lucas-zimerman) [reply src/Sentry/Internal/MainSentryEventProcessor.cs:116]: ty for pointing out.

File: test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs (hunk 5 above)
Comment (lucas-zimerman) [reply test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs:84]: Just wondering, 
https://github.com/getsentry/sentry-dotnet/blob/4c6fdac57a5cd8179d0af6fb8e219c56767a2c90/test/Sentry.Tests/Internals/MainSentryEventProcessorTests.cs#L104

`Process_SendDefaultPiiTrueIdEnvironmentFalse_UserNameNotSet
`
Should become 
`Process_SendDefaultPiiTrueIdAndEnvironmentFalse_UserNameNotSet `
?

**Comment by lucas-zimerman** [comment]
I added the conflicting changelog from unreleased to vNext, If needed I'll move again to unreleased.

**Review by bruno-garcia — APPROVED** [review]
Thanks for another contribution @lucas-zimerman !

**Pull request merged**